discovery/watcher.go

```diff
@@ -488,85 +474,82 @@ func (w *Watcher) getDataplaneFeatures() (map[string]bool, error) {
 //
-// This returns an addrSet containing the most recent set of servers. If the
-// addrSet is nil, then there were not changes seen to the server set. (This is
-// the case if the server watch stream is disabled.
-//
-// When this returns with an error, the current server should no longer be
-// considered "OK". This may return a non-nil addrSet and a non-nil error (it
-// will usually do this when the server watch stream is aborted).
-func (w *Watcher) watch() (*addrSet, error) {
+// This updates addrs in place to add or remove servers found from the server
+// watch stream, if applicable.
+func (w *Watcher) watch(addrs *addrSet) error {
 	current := w.currentServer.Load().(serverState)
 	if current.dataplaneFeatures["DATAPLANE_FEATURES_WATCH_SERVERS"] && !w.config.ServerWatchDisabled {
-		return w.watchStream()
+		return w.watchStream(addrs)
 	} else {
 		return w.watchSleep()
 	}
 }
 
 // watchStream opens a gRPC stream to receive server set changes. This blocks
-// potentially forever.
+// potentially forever. It updates addrs in place, adding or removing servers
+// to match the response from the server watch stream.
 //
 // This may be aborted when the gRPC stream receives some error, when the
-// Watcher's context is cancelled, or (TODO) when the Watcher is told to switch
+// Watcher's context is cancelled, or when the Watcher is told to switch
 // servers.
-//
-// If an error that aborts the gRPC stream, we pass the non-nil error back,
-// and the server is marked unhealthy elsewhere.
-func (w *Watcher) watchStream() (*addrSet, error) {
+func (w *Watcher) watchStream(addrs *addrSet) error {
 	w.log.Debug("Watcher.watchStream")
 	client := pbserverdiscovery.NewServerDiscoveryServiceClient(w.conn)
 	serverStream, err := client.WatchServers(w.ctxForSwitch, &pbserverdiscovery.WatchServersRequest{})
 	if err != nil {
 		w.log.Error("unable to open server watch stream", "error", err)
-		return nil, err
+		return err
 	}
 
-	var set *addrSet
 	for {
 		// This blocks until there is a change from the server.
 		resp, err := serverStream.Recv()
 		if err != nil {
 			if !errors.Is(err, context.Canceled) {
 				w.log.Error("unable to receive from server watch stream", "error", err)
 			}
-			return set, err
+			return err
 		}
 
-		// The set of servers from the stream is the best known set of servers to use.
-		set = newAddrSet()
+		// Collect addresses from from the stream.
+		streamAddrs := []Addr{}
 		for _, srv := range resp.Servers {
 			addr, err := w.nodeToAddrFn(srv.Id, srv.Address)
 			if err != nil {
 				// ignore the server on failure.
 				w.log.Warn("failed to parse server address from watch stream", "error", err)
 				continue
 			}
-			set.Put(OK, addr)
+			streamAddrs = append(streamAddrs, addr)
 		}
+
+		// Update the addrSet. This sets the lastUpdated timestamp for each address,
+		// and removes servers not present in the server watch stream.
+		addrs.SetAddrs(streamAddrs...)
```

Comment:
Yeah that's true, but I think in that case the offending server wouldn't consider itself unhealthy and push an update on the stream to remove itself - so when the clients eventually figure out that something is wrong (e.g. via missed keepalive ACKs) they'd just move on to the next server.

I guess it could still be a problem if whatever server they land on is already oversubscribed, though?

I'm also wondering whether there's a potential failure-mode where the server you're connected to is so overwhelmed it's unable to process memberlist and autopilot traffic, so tells you that it is the only healthy server. I guess in that case, falling back to DNS/script resolution would fix it - sorry, just thinking aloud!

Reply:
I merged this, but making sure to conclude this discussion.

Basically, my thought is probably not a big concern because:

1. consul-dataplanes would only switch servers when they are disconnected, and servers flapping in the server watch stream would not immediately cause them to be disconnected
2. consul-dataplane should not forget about the server it is currently connected to, because that server would not flap in the server watch stream, because a server would not consider itself unhealthy.
3. consul-dataplane will try _different_ server on its next attempt, because it should always remember it attempted the server it was most-recently connected to, because of (2), even if it has forgotten all other servers